src/Features/Core/Portable/CodeRefactorings/CodeRefactoringService.cs

```diff
@@ -30,43 +30,73 @@ namespace Microsoft.CodeAnalysis.CodeRefactorings;
+                    // if values do not match enum TextDocumentKind. Here we'll throw too.
+                    var kind = (TextDocumentKind)Enum.Parse(typeof(TextDocumentKind), documentKind, ignoreCase: true);
+                    var documentExtensions = provider.Metadata.DocumentExtensions.Where(e => !string.IsNullOrEmpty(e));
+                    if (!documentExtensions.Any())
+                    {
+                        // Metadata without file extension applies to all files.
+                        documentExtensions = [""];
+                    }
+
+                    foreach (var documentExtension in documentExtensions)
+                    {
+                        var orderable = new OrderableLanguageDocumentMetadata(
+                            provider.Metadata.Name ?? "", language, kind, documentExtension, provider.Metadata.AfterTyped, provider.Metadata.BeforeTyped);
+                        yield return new Lazy<CodeRefactoringProvider, OrderableLanguageDocumentMetadata>(() => provider.Value, orderable);
+                    }
+                }
             }
         }
     }
 
-    private ImmutableDictionary<string, Lazy<ImmutableArray<CodeRefactoringProvider>>> LanguageToProvidersMap
-        => _lazyLanguageToProvidersMap.Value;
+    private ImmutableDictionary<ProviderKey, Lazy<ImmutableArray<CodeRefactoringProvider>>> LanguageDocumentToProvidersMap
+        => _lazyLanguageDocumentToProvidersMap.Value;
 
     private ImmutableDictionary<CodeRefactoringProvider, CodeChangeProviderMetadata> RefactoringToMetadataMap
         => _lazyRefactoringToMetadataMap.Value;
 
-    private ConcatImmutableArray<CodeRefactoringProvider> GetProviders(TextDocument document)
+    internal ConcatImmutableArray<CodeRefactoringProvider> GetProviders(TextDocument document)
     {
         var allRefactorings = ImmutableArray<CodeRefactoringProvider>.Empty;
-        if (LanguageToProvidersMap.TryGetValue(document.Project.Language, out var lazyProviders))
+
+        // Include providers which apply to all extensions
+        var key = new ProviderKey(document.Project.Language, document.Kind, "");
+        if (LanguageDocumentToProvidersMap.TryGetValue(key, out var lazyProviders))
+        {
+            allRefactorings = lazyProviders.Value;
+        }
+
+        // Get providers for specific combination of language, doc kind and extension,
+        // e.g. (C#, AdditoinalDocument, .xaml)
+        if (FileNameUtilities.GetExtension(document.FilePath) is string documentExtension && documentExtension.Length > 0)
         {
-            allRefactorings = ProjectCodeRefactoringProvider.FilterExtensions(document, lazyProviders.Value, GetExtensionInfo);
+            key = new ProviderKey(document.Project.Language, document.Kind, documentExtension);
+            if (LanguageDocumentToProvidersMap.TryGetValue(key, out lazyProviders))
+            {
+                allRefactorings = allRefactorings.Concat(lazyProviders.Value);
```

Comment:
Having this extra "concat" here means we're losing the allocation-freeness of the ConcatFast being used later. Not sure how much that matters?

Reply:
I am open to suggestions :)